core/src/main/java/com/linecorp/armeria/common/multipart/Multipart.java

```diff
@@ -361,4 +367,45 @@ static Multipart from(String boundary, Publisher<? extends HttpData> contents, B
      */
     CompletableFuture<AggregatedMultipart> aggregateWithPooledObjects(EventExecutor executor,
                                                                       ByteBufAllocator alloc);
+
+    /**
+     * Collects the elements published by this {@link Multipart}.
+     * The returned {@link CompletableFuture} will be notified when the elements are fully consumed.
+     *
+     * <p>Note that if this {@link Multipart} was subscribed by other {@link Subscriber} already,
+     * the returned {@link CompletableFuture} will be completed with an {@link IllegalStateException}.
+     *
+     * @param function A {@link Function} that processes the {@link BodyPart} and returns a future that will
+     *                 complete with the process result. The {@link Function} must consume the {@link BodyPart}.
+     *                 And If not, collect method will stop processing next {@link BodyPart}.
+     * <p>For example:
+     * <pre>{@code
+     * Path tempDir = ...;
+     * CompletableFuture<List<Object>> collect =
+     *         Multipart.of(request)
+     *                  .collect(bodyPart -> {
+     *                      if (bodyPart.filename() != null) {
+     *                          final Path path = tempDir.resolve(bodyPart.name());
+     *                          return bodyPart.writeTo(path).thenApply(ignore -> path);
+     *                      }
+     *                      return bodyPart.aggregate().thenApply(AggregatedHttpObject::contentUtf8);
+     *                  });
+     * }</pre>
+     */
+    @UnstableApi
+    default <T> CompletableFuture<List<T>> collect(
+            Function<? super BodyPart, CompletableFuture<? extends T>> function) {
+        return collect(function, EMPTY_OPTIONS);
+    }
+
+    /**
+     * Collects the elements published by this {@link Multipart}.
+     * The returned {@link CompletableFuture} will be notified when the elements are fully consumed.
+     *
+     * <p>Note that if this {@link Multipart} was subscribed by other {@link Subscriber} already,
+     * the returned {@link CompletableFuture} will be completed with an {@link IllegalStateException}.
+     */
+    @UnstableApi
+    <T> CompletableFuture<List<T>> collect(Function<? super BodyPart, CompletableFuture<? extends T>> function,
+                                           SubscriptionOption... options);
```

Comment:
Recently, thanks to @sleeplesslord, `StreamMessage.mapAsync()` that was added which might be perfect for this. #3962

```suggestion
    default <T> CompletableFuture<List<T>> collect(Function<? super BodyPart, CompletableFuture<? extends T>> function,
                                           SubscriptionOption... options) {
        return bodyParts().mapAsync(function).collect(options);                                
   }
```